Source/Src_nd/update_scal.F90

```diff
@@ -14,514 +14,248 @@ module update_scal_module
-                end do
-             end if
-          enddo
-       enddo
-    enddo
-
-  end subroutine update_rhoX_3d
+integer         , intent(in   ) :: z_lo(3), z_hi(3)
+double precision, intent(in   ) :: sfluxz(z_lo(1):z_hi(1),z_lo(2):z_hi(2),z_lo(3):z_hi(3),nscal)
 #endif
-
-
-#if (AMREX_SPACEDIM == 1)
-  subroutine update_rhoh_1d(lev, lo, hi, &
-       sold,   so_lo, so_hi, nc_so, &
-       snew,   sn_lo, sn_hi, nc_sn, &
-       sfluxx, x_lo, x_hi, nc_x, &
-       force,  f_lo, f_hi, nc_f, &
-       p0_new, &
-       dx, dt, &
-       nspec) bind(C,name="update_rhoh_1d")
-
-    integer         , intent(in   ) :: lev, lo(1), hi(1)
-    integer         , intent(in   ) :: so_lo(1), so_hi(1), nc_so
-    double precision, intent(in   ) :: sold  (so_lo(1):so_hi(1),nc_so)
-    integer         , intent(in   ) :: sn_lo(1), sn_hi(1), nc_sn
-    double precision, intent(inout) :: snew  (sn_lo(1):sn_hi(1),nc_sn)
-    integer         , intent(in   ) :: x_lo(1), x_hi(1), nc_x
-    double precision, intent(in   ) :: sfluxx(x_lo(1):x_hi(1),nc_x)
-    integer         , intent(in   ) :: f_lo(1), f_hi(1), nc_f
-    double precision, intent(in   ) :: force (f_lo(1):f_hi(1),nc_f)
-    double precision, intent(in   ) :: p0_new(0:max_radial_level,0:nr_fine-1)
-    double precision, intent(in   ) :: dx(1), dt
-    integer         , intent(in   ) :: nspec
-
-    ! Local variables
-    integer          :: i
-    double precision :: divterm
-
-    integer :: pt_index(3)
-    type(eos_t)      :: eos_state
-
-    do i=lo(1),hi(1)
-       divterm = (sfluxx(i+1,rhoh_comp) - sfluxx(i,rhoh_comp))/dx(1)
-       snew(i,rhoh_comp) = sold(i,rhoh_comp) + dt*(-divterm + force(i,rhoh_comp))
-    end do
-
-    if (do_eos_h_above_cutoff) then
-       do i = lo(1), hi(1)
-          if (snew(i,rho_comp) .le. base_cutoff_density) then
-
-             eos_state%rho = snew(i,rho_comp)
-             eos_state%T   = sold(i,temp_comp)
-             eos_state%p   = p0_new(lev,i)
-             eos_state%xn  = snew(i,spec_comp:spec_comp+nspec-1)/eos_state%rho
-
-             pt_index(:) = (/i, -1, -1/)
-
-             ! (rho,P) --> T,h
-             call eos(eos_input_rp, eos_state, pt_index)
-
-             snew(i,rhoh_comp) = snew(i,rho_comp) * eos_state%h
-
-          end if
-       enddo
-    end if
-
-  end subroutine update_rhoh_1d
+integer         , intent(in   ) :: f_lo(3), f_hi(3)
+double precision, intent(in   ) :: force (f_lo(1):f_hi(1),f_lo(2):f_hi(2),f_lo(3):f_hi(3),nscal)
+double precision, intent(in   ) :: dx(3)
+double precision, value, intent(in   ) :: dt
+integer  , value, intent(in   ) :: startcomp, endcomp
+
+! Local variables
+integer          :: i, j, k, comp, comp2
+double precision :: divterm
+double precision :: delta, frac, sumX
+logical          :: has_negative_species
+
+!$gpu
+
+do comp = startcomp, endcomp
+   do k = lo(3), hi(3)
+      do j = lo(2), hi(2)
+         do i = lo(1), hi(1)
+
+            divterm = (sfluxx(i+1,j,k,comp) - sfluxx(i,j,k,comp))/dx(1)
```

Comment:
I notice that the 3D version doesn't have the "has_negative_species_ logic" so this may change results in 1D/2D (just as an FYI).